### ISSUE TITLE
AJV 6.5.2 is not IE11 compatible as it still depends on uri-js version 4.2.1 instead of 4.2.2

### ISSUE THREAD

**Issue by mduu** [ajv-validator/ajv#833]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**

4.5.2 / yes

**The problem**

Ajv 4.5.2 depends on uri-js version `^4.2.1`. This version of uri-js had the problem that it forced packers to use the "esnext" dist instead - as previous versions - the es5 version. This lead into breaking IE11 compatibility. Regarding the uri-js project they fixed the issue with uri-js version 4.2.2. Ajv still uses depends on the broken version 4.2.1 even if they released 4.2.2 some months ago.

So after upgrading our projects to the latest AJV (4.5.2) it crashes in IE 11 because esnext-code of uri-js gets picked up.

See [this issue](https://github.com/garycourt/uri-js/issues/24) over at uri-js for more information. Especially the last 2-3 comments.

**Expected**

Expect to get the ES5 code of uri-js which should be true if AJV depends on the 4.2.2 patch of uri-js. If AJV claims to be IE11 / ES5 compatible it should fix this dependency so the workaround below is not required to use AJV with ES5.

**Workaround**

In the app's package.json one can add a dependency to uri-js version ^4.2.2 which will then force NPM/yarn to use the patched version of uri.js. But this is a workaround as the app itself does not require this library.

**Comment by epoberezkin** [comment]
@mduu The dependency version is "^4.2.1" that means that the latest "4.x.x" version will be installed - it is currently 4.2.2. Why it is an issue and why #834 is needed?

**Comment by mduu** [comment]
@epoberezkin Very good question. I'm not really a NPM/yarn expert but Yarn (and npm) both installed 4.2.1 of uri-js instead the latest one. I know that Nuget would do it like you described if it is only the third number (others will not auto-upgrade). When many packages have dependency on uri-js package-manager normally try to negotiate the version that matches best for all. This may be not the latest one but something in between. Not sure about this with NPM/yarn.

As we know that AJV has problems with 4.2.1 I suggest to upgrade its dependency to a version we know should be "safe" for AJV. This way AJV can guarantee it will work properly (even with ES5).

**Comment by epoberezkin** [comment]
> Yarn (and npm) both installed 4.2.1 of uri-js instead the latest one

Does it happen when you install from empty node_modules folder? The problem may be with your npm setup as well.

> When many packages have dependency on uri-js package-manager normally try to negotiate the version that matches best for all

Where is this behaviour described? My understanding is that if the latest version is 4.2.2 than it is the only one that matches “^4.2.1”

**Comment by mduu** [comment]
>Does it happen when you install from empty node_modules folder? The problem may be with your npm setup as well.

I got 4.2.1 on a clean `node_modules` directory.

>Where is this behaviour described? My understanding is that if the latest version is 4.2.2 than it is the only one that matches “^4.2.1”

I searched the web quite a while but could not find any documentation how NPM negotiate the final version number.

But thinking of it its relatively clear that it will not just use the latest patch but it needs to check the dependencies of other packages to the same target package. Eg. ajv says ^4.2.1 but another package I have in package.json may say "4.2.1". Then I definitively get 4.2.1 which breaks things. If ajv knows that 4.2.1 results in problem and therefore specify ^4.2.2 then the above situation results in a error the developers have to check and solve manually which is correct.
